Customizing/global/plugins/Services/Cron/CronHook/UserOrguImport/classes/OrgusCleanup.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace CaT\IliasUserOrguImport;
+
+class OrgusCleanup
+{
+
+	protected $tree;
+	protected $rep_utils;
+	protected $grm;
+	protected $oc;
+
+	public function __construct($tree, $rep_utils, IliasGlobalRoleManagement $grm, Orgu\OrguConfig $oc)
+	{
+		$this->tree = $tree;
+		$this->rep_utils = $rep_utils;
+		$this->grm = $grm;
+		$this->oc = $oc;
+	}
+
+	/**
+	 * Delete all import subtree orgus, which have no user assignments and only
+	 * contain orgus without user assignments.
+	 *
+	 * @return void
+	 */
+	public function cleanupEmptyOrgus()
+	{
+		$exit_subtree = $this->tree->getSubTree(
+			$this->tree->getNodeTreeData($this->oc->getExitRefId()),
+			false,
+			'orgu'
+		);
+		$root_ref_id = $this->oc->getRootRefId();
+		$delete_candidates = [];
+		// all leafs are candidates, since they obviously contain no
+		// orgus which have any assignment, so get all leafs first
+		foreach ($this->tree->getSubTree($this->tree->getNodeTreeData($root_ref_id), false, 'orgu') as $ref_id) {
+			if ($ref_id !== $root_ref_id && !in_array($ref_id, $exit_subtree)) {
+				// leafs contain no children
+				if (count($this->tree->getChildsByType($ref_id, 'orgu')) === 0) {
+					$delete_candidates[] = (int)$ref_id;
+				}
+			}
+		}
+		$delete = [];
+		while ($candidate = array_shift($delete_candidates)) {
+			$orgu = new \ilObjOrgUnit($candidate);
+			// a candidate should be deleted, if it has no user assignments
+			if (0 === $this->grm->numberOfAssignedUsers([$orgu->getSuperiorRole(),$orgu->getEmployeeRole()])) {
+				$delete[] = $candidate;
+				// if an orgu is to be deleted, its parent is a delete candidate as well
+				$parent = (int)$this->tree->getParentId($candidate);
+				if ($parent !== $root_ref_id) { // ignore tree root
+					array_push($delete_candidates, $parent);
+				}
+			}
+		}
+		$this->remove($delete);
+	}
+
+	/**
+	 * Remove orgus with provided information.
+	 *
+	 * @param	int[]	$ref_ids
+	 * @return	void
+	 */
+	public function remove(array $ref_ids)
+	{
+		$to_delete_rec = [];
+		foreach ($ref_ids as $ref_id) {
+			foreach ($this->tree->getSubTree($this->tree->getNodeTreeData($ref_id), false, 'orgu') as $c_ref_id) {
+				if (!in_array((int)$c_ref_id, $to_delete_rec)) {
+					$to_delete_rec[] = (int)$c_ref_id;
+				}
+			}
+		}
+		if (count($to_delete_rec) > 0) {
+			$to_delete = [];
+			foreach ($to_delete_rec as $ref_id) {
+				$parent_id = (int)$this->tree->getParentId($ref_id);
```

Comment:
Warum wird hier auch noch die Elterneinheit gelöscht?

Reply:
Wird sie nicht, $to_delete[] = $ref_id; nicht $to_delete[] = $parent_id;
Hier geht es darum, nur die top-orgus des Unterbaumes zu finden. Die suborgus werden successive automatisch mitentfernt.

Reply:
Vielleicht geht das gerade nicht in meinen Kopf. Aber wann wird die Ausgangsorgeinheit gelöscht, wenn deren Parent nicht NULL ist, aber noch Mitglieder hat?

A = 2 Member
 AA = 2 Member
   AAA = 0 Member
    AAAA = 0 Member

Wie ich das verstanden habe, wird jetzt nur AAAA gelöscht. Bzw in dem Beispiel sogar gar keine.

Reply:
AAA wird gelöscht (oder sollte gelöscht werden). Die RepUtils kümmern sich dann wohl automatisch darum, dass AAAA mit verschwinden. In dieser Methode kommen nur einheiten an, die forher als rekursiv leer gefunden wurden. Dann suche ich diejenigen, die die Wurzeln dieses Unterbaums darstellen. Das sind genau die, deren eltern nicht im Unterbaum liegen. Die gebe ich dann an RepUtils, welches damit den gsamten baum löscht.

Reply:
Dann versteh ich es nicht. 
Laut meinen Kopf, würde aus Zeile 62 AAA und AAAA zurück kommen. Weil die einen cnt von 0 haben. Dann suchst du in remove nach deren Subtree und willst den löschen. AAA und AAAA kommen aber nie in $to_delete an.

Reply:
Nach deinem Beispiel ist das Argument dieser methode ref_ids = [AAA, AAAA].
AAAA.parent = AAA
AAA.parent = AA
nun ist AA nicht in ref_ids, also schließe ich daraus, dass AAA die Wurzel des zu Löschenden Unterbaumes
AAA
AAAA
ist und gebe damit nur AAA an RepUtils

Reply:
Ok das scheint zu klappen, weil getSubTree sich selber als zurück gibt. Also AAA->getSubTree() beinhaltet auch AAA.